var/spack/repos/builtin/packages/infer/package.py

```diff
@@ -0,0 +1,118 @@
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+import re
+
+
+class Infer(Package):
+    """Infer: a static analyzer for Java, C, C++, and Objective-C
+
+       Infer is a static program analyzer for Java, C, and
+       Objective-C, written in OCaml. Infer is deployed within
+       Facebook and it is running continuously to verify select
+       properties of every code modification for the main Facebook
+       apps for Android and iOS, Facebook Messenger, Instagram, and
+       other apps. It can be used for other code too: Infer can also
+       analyze C code, and Java code that is not Android.  At present
+       Infer is tracking problems caused by null pointer dereferences
+       and resource and memory leaks, which cause some of the more
+       important problems on mobile. """
+
+    homepage = "http://fbinfer.com/"
+    url      = "https://github.com/facebook/infer/archive/v0.13.0.tar.gz"
+
+    version('develop', git='https://github.com/facebook/infer', branch='master')
+    # Latest stable version currently does not build due to opam
+    # package version conflict; see
+    # https://github.com/facebook/infer/issues/830
+    # version('0.13.0', 'd6e18dbcfa846feb1547bff4dbdd96f0')
+
+    depends_on('autoconf', type=('build'))
+    depends_on('automake', type=('build'))
+    depends_on('cmake', type=('build'))
+    depends_on('libtool', type=('build'))
+    depends_on('ocaml', type=('build'))
+    depends_on('opam', type=('build'))
+    depends_on('pkg-config', type=('build'))
+
+    parallel = False
+
+    def setup_environment(self, spack_env, run_env):
+        spack_env.set('OPAMYES', '1')
+        spack_env.set('INFER_CONFIGURE_OPTS',
+                      ('--prefix={0}'.format(self.prefix) +
+                       ' --disable-ocaml-binannot' +
+                       ' --without-fcp-clang'))
+
+    def install(self, spec, prefix):
+        # Install a local version of opam for building Infer. Options
+        # taken from
+        # https://github.com/Homebrew/homebrew-core/blob/master/Formula/infer.rb
+        opamroot = join_path('.', 'opamroot')
+        mkdirp(opamroot)
+        env['OPAMROOT'] = opamroot
+        opam = Executable(join_path(self.spec['opam'].prefix.bin, 'opam'))
+        opam('init', '--no-setup')
+
+        # Detect OCaml version used in Infer build scropt
+        build_file_path = join_path('.', 'build-infer.sh')
+        f = open(build_file_path, 'r')
```

Comment:
The idiomatic way to do something like this is:
```python
with open(build_file_path, 'r') as f:
     pattern = re.compile('OCAML_VERSION=\${OCAML_VERSION:-\"([^\"]+)\"}')
     ...
```
Using a context manager is basically the accepted RAII idiom for python, and be ensured you'll always call `f.close()` (even in exceptional paths).

Reply:
Done.